contracts/src/UniswapV3MultiQuoter.sol

```diff
@@ -166,22 +167,39 @@ contract UniswapV3MultiQuoter {
                 zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1
             );
 
-            for (uint256 i = 0; i < amountsOut.length; ++i) {
-                amountsOut[i] = zeroForOne ? uint256(result.amounts0[i]) : uint256(result.amounts1[i]);
-                gasEstimate[i] += result.gasEstimates[i];
+            for (uint256 i = 0; i < nextAmountsLength; ++i) {
+                uint256 amountReceived = zeroForOne ? uint256(-result.amounts1[i]) : uint256(-result.amounts0[i]);
+                if (amountReceived != amountsOut[i]) {
+                    // for exact output swaps we need to check whether we would receive the full amount due to
+                    // multiswap behavior when hitting the limit
```

Comment:
is this logic something we can remove if we fix the behavior in multiswap? maybe that would be a better place to fix it instead of here.

Reply:
I wanted to separate out the logic for multiswap since the logic there is already very complex. This special handling only needs to be done for exact output swaps so I thought this was a better area for this logic. I think multiswap should simply return deltas of of token0 and token1. Happy to discuss further offline.

Reply:
understood.
just to clarify what this logic is doing -- I see in line 193 it's zeroing out the amountsIn[i] if Multiswap wasn't able to fullfill the whole amountOut[i]. Is that to prevent the consumer (router?) of these amountsIn from breaking?

Reply:
Ah actually that is not needed since `amountsIn` will be set to 0 when initialized. 
Just a rough overview of this logic:
1. Check whether `multiswap` result returns amount to be received matches the amount we **want** out
2. If result has the full amount to be received, populate the `amountsOut` for the next hop based on amount payable 
3. If result doesn't have the full amount to be received, set a variable to track how many valid output amounts we have that we can send into the next hop (`nextAmountsLength`)
4. Repeat multiswap for each hop using a new amounts array with valid output amounts from previous hop